### PR TITLE
Feature/andrembcosta/multi resolution solver

### PR DESCRIPTION
**Pull request by andrembcosta**
Creating PR for Multi-Resolution solver

### CODE REVIEW

**Comment by CusiniM** [comment]
@andrembcosta I think you deleted the `integratedTests` directory...

**Comment by andrembcosta** [comment]
Should I re-create it? I think I was having some problems authenticating in the integratedTests repository.

**Comment by CusiniM** [comment]
> Should I re-create it? I think I was having some problems authenticating in the integratedTests repository.

I think I have fixed it but if you have permissions problems let's solver them. Deleting it seems to create quite a few headaches when merging.

**Comment by CusiniM** [comment]
fracturedElements should not be empty anymore. I will fix the external set later.

File: src/coreComponents/physicsSolvers/multiphysics/MultiResolutionHFSolver.cpp [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+}
+
+//in the initial step and whenever the subproblem mesh moves, we must update the node map
+// void MultiResolutionHFSolver::updateNodeMaps( MeshLevel const & base,
+//            MeshLevel const & patch )
+// {
+//   // get list of nodes on the boundary of the patch
+//   NodeManager const & patchNodeManager = patch.getNodeManager();
+//   NodeManager const & baseNodeManager = base.getNodeManager();
+//   SortedArrayView< localIndex const > const patchExternalSet = patchNodeManager.externalSet();
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/multiphysics/MultiResolutionHFSolver.cpp:193]: I have filled the `m_isExternal` in the `nodeManager`. I think you need to use that one and not this `externalSet()`.

**Comment by andrembcosta** [comment]
@rrsettgast, @CusiniM 

what about merging this simplified version where the meshes of base and patch are identical?

**Comment by andrembcosta** [comment]
To do:
- there seems to be a type issue with a 2D array in GPU builds;
- verify correct path to apply BCs with multiple bodies;
- add MR integratedTest and rebaseline PF tests because of change in fieldName.

**Comment by andrembcosta** [comment]
Issue with objectPath in the inputFile.

In the MR solver, we need to specify the displacement BCs to the base problem. Prior to this most recent develop merge, the following FieldSpecification was doing that:

`<FieldSpecification`
      `name="ypos"`
      `objectPath="base/Level0/nodeManager"`
      `fieldName="totalDisplacement"`
      `component="1"`
      `scale="1.0e-7"`
      `functionName="f_b"`
      `setNames="{ ypos }"/>`

now, seem that it is not finding this objectPath. What worked was to use `objectPath="nodeManager"`, which seems to default to MeshBody(0) and give the right answer.

File: src/coreComponents/mesh/simpleGeometricObjects/BoundedPlane.cpp [hunk 2]
```diff
@@ -61,6 +61,32 @@ BoundedPlane::BoundedPlane( const string & name, Group * const parent ):
...
+//constructor given two points, used for 2.5D problems
+BoundedPlane::BoundedPlane( const real64 oldX, const real64 oldY,
+                            const real64 newX, const real64 newY,
+                            const string & name, Group * const parent ):
+  SimpleGeometricObjectBase( name, parent ),
+  m_origin{ 0.0, 0.0, 0.0 },
+  m_normal{ 0.0, 0.0, 1.0 },
+  m_lengthVector{ 0.0, 0.0, 0.0 },
+  m_widthVector{ 0.0, 0.0, 0.0 },
+  m_tolerance()
```
Comment (CusiniM) [inline src/coreComponents/mesh/simpleGeometricObjects/BoundedPlane.cpp:73]:
```suggestion
  m_tolerance(1.0e-5)
```

File: src/coreComponents/mesh/simpleGeometricObjects/BoundedPlane.cpp [hunk 3]
```diff
@@ -61,6 +61,32 @@ BoundedPlane::BoundedPlane( const string & name, Group * const parent ):
...
+{
+  m_origin = { (oldX + newX)/2.0, (oldY + newY)/2.0, 0.0};
+  m_normal = { -(newY-oldY), newX-oldX, 0.0 };
+  m_lengthVector = { newX-oldX, newY-oldY, 0.0 };
+  m_widthVector = { 0.0, 0.0, 1.0 };
+  real64 norm = std::sqrt( pow( newX-oldX, 2 )+pow( newY-oldY, 2 ));
+  m_dimensions.resize( 2 );
+  m_dimensions[0] = norm+1e-4;
```
Comment (CusiniM) [inline src/coreComponents/mesh/simpleGeometricObjects/BoundedPlane.cpp:81]: Can you add some comments and some TODOs? These ad hoc tolerances seem a bit hacky. I think it's fine at this stage but it would be good to clean things up at some point.

File: src/coreComponents/mesh/simpleGeometricObjects/BoundedPlane.hpp [hunk 4]
```diff
@@ -45,6 +45,14 @@ class BoundedPlane : public SimpleGeometricObjectBase
   BoundedPlane( const string & name,
                 Group * const parent );
 
+  /**
+   * @brief Internal constructor.
+   * @param name name of the object in the data hierarchy.
+   * @param parent pointer to the parent group in the data hierarchy.
+   */
```
Comment (CusiniM) [inline src/coreComponents/mesh/simpleGeometricObjects/BoundedPlane.hpp:52]: add comment about how this constructor is used.

File: src/coreComponents/mesh/NodeManager.cpp [hunk 5]
```diff
@@ -105,6 +105,44 @@ void NodeManager::buildSets( CellBlockManagerABC const & cellBlockManager,
...
+  ArrayOfArraysView< localIndex const > const & faceToNodes = faceManager.nodeList().toViewConst();
+
+  // get the "isExternal" field from for *this, and set it to zero
+  m_isExternal.zero();
+
+  // get externalSet reference to insert external faces
+  SortedArray< localIndex > & externalSet = this->externalSet();
+
+  //get normals to all faces
+  const arrayView2d< real64 const > normalOfAllFaces = faceManager.faceNormal();
```
Comment (CusiniM) [inline src/coreComponents/mesh/NodeManager.cpp:123]:
```suggestion
 arrayView2d< real64 const >    const  normalOfAllFaces = faceManager.faceNormal();
```

File: src/coreComponents/mesh/NodeManager.cpp [hunk 6]
```diff
@@ -105,6 +105,44 @@ void NodeManager::buildSets( CellBlockManagerABC const & cellBlockManager,
...
+
+  // loop through all faces
+  for( localIndex kf=0; kf<faceManager.size(); ++kf )
```
Comment (CusiniM) [inline src/coreComponents/mesh/NodeManager.cpp:126]: this should be a `forAll<parallelHostPolicy>`. careful not to capture member variables though.

File: src/coreComponents/mesh/NodeManager.cpp [hunk 7]
```diff
@@ -105,6 +105,44 @@ void NodeManager::buildSets( CellBlockManagerABC const & cellBlockManager,
   } );
 }
 
+void NodeManager::setIsExternal( FaceManager const & faceManager )
```
Comment (CusiniM) [inline src/coreComponents/mesh/NodeManager.cpp:108]: what is the difference between this method and `NodeManager::setDomainBoundaryObjects`?  I may not understand what external mean here coz it sounds to me the same as being on the boundary.

File: src/coreComponents/physicsSolvers/multiphysics/MultiResolutionHFSolver.cpp [hunk 8]
```diff
@@ -0,0 +1,567 @@
...
+//  if (dist < tolerance)
+//    {
+//      return a;
+//    }
+//     }
+//   return -1;
+// }
+
+//this function checks if an index is already in an array
+bool isMember( localIndex const & ind, array1d< localIndex > const & arr )
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/multiphysics/MultiResolutionHFSolver.cpp:71]: this should go in LvArray if really necessary. Otherwise, would this work?

if( `std::find(arr.begin(), arr.end(), ind) != arr.end() `)

File: src/coreComponents/physicsSolvers/multiphysics/MultiResolutionHFSolver.hpp [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+  string m_baseSolverName;
+
+  //name of patch phase-field fracture solver
+  string m_patchSolverName;
+
+  // list of damage dofs to be fixed in the subdomain boundary
+  array1d< localIndex > m_nodeFixDamage;
+
+  // list of disp dofs to be fixed in the subdomain boundary
+  array1d< localIndex > m_nodeFixDisp;
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/multiphysics/MultiResolutionHFSolver.hpp:123]: I am not sure you are using the right type of containers for these. Why don't you use sets (`sortedArray`) instead of arrays? I think it would make sense since you are always checking if they are unique. Maybe they can even jsut be std objects?

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp [hunk 10]
```diff
@@ -1308,5 +1333,100 @@ SolidMechanicsLagrangianFEM::scalingForSystemSolution( DomainPartition const & d
   return 1.0;
 }
 
+void SolidMechanicsLagrangianFEM::setInternalBoundaryConditions( array1d< localIndex > const & fixedNodes,
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp:1336]:
```suggestion
void SolidMechanicsLagrangianFEM::setInternalBoundaryConditions( arrayView1d< localIndex const > const & fixedNodes,
```

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp [hunk 11]
```diff
@@ -1308,5 +1333,100 @@ SolidMechanicsLagrangianFEM::scalingForSystemSolution( DomainPartition const & d
...
+                                                                 array2d< real64 > const & fixedValues )
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp:1337]:
```suggestion
                                                                 arrayView2d< real64 const > const & fixedValues )
```

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp [hunk 12]
```diff
@@ -1308,5 +1333,100 @@ SolidMechanicsLagrangianFEM::scalingForSystemSolution( DomainPartition const & d
...
+{
+  localIndex count = 0;
+  m_fixedDisplacementNodes.resize( fixedNodes.size() );
+  m_fixedDisplacementValues.resize( fixedNodes.size(), 3 );
+  for( localIndex c : fixedNodes )
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp:1342]: sould probably use a forAll here.

File: src/coreComponents/physicsSolvers/surfaceGeneration/EmbeddedSurfaceGenerator.cpp [hunk 13]
```diff
@@ -109,123 +109,354 @@ void EmbeddedSurfaceGenerator::initializePostSubGroups()
...
   // Get meshLevel
-  MeshLevel & meshLevel = domain.getMeshBody( 0 ).getBaseDiscretization();
-
-  // Get managers
-  ElementRegionManager & elemManager = meshLevel.getElemManager();
-  NodeManager & nodeManager = meshLevel.getNodeManager();
-  EmbeddedSurfaceNodeManager & embSurfNodeManager = meshLevel.getEmbSurfNodeManager();
-  EdgeManager & edgeManager = meshLevel.getEdgeManager();
-  arrayView2d< real64 const, nodes::REFERENCE_POSITION_USD > const & nodesCoord = nodeManager.referencePosition();
+  forDiscretizationOnMeshTargets( domain.getMeshBodies(), [&] ( string const &,
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/surfaceGeneration/EmbeddedSurfaceGenerator.cpp:112]: is this so that you only create the embedded fracture on one of the 2 meshes?

File: src/coreComponents/physicsSolvers/surfaceGeneration/EmbeddedSurfaceGenerator.cpp [hunk 14]
```diff
@@ -109,123 +109,354 @@ void EmbeddedSurfaceGenerator::initializePostSubGroups()
...
+
+    KERNEL_TYPE::template launchCIComputationKernel< parallelDevicePolicy< 32 >, KERNEL_TYPE >( kernel );
+  } );
+
 
   // add all new nodes to newObject list
-  for( localIndex ni = 0; ni < embSurfNodeManager.size(); ni++ )
+  // also, get index of edges that were just cut
+  array1d< globalIndex > newEdges;
+  array1d< globalIndex > & parentEdgeGlobalIndex = embSurfNodeManager.getParentEdgeGlobalIndex();
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/surfaceGeneration/EmbeddedSurfaceGenerator.cpp:394]: you don't seem to be resizing this. You should get a view probably.

File: src/coreComponents/physicsSolvers/surfaceGeneration/EmbeddedSurfaceGenerator.cpp [hunk 15]
```diff
@@ -109,123 +109,354 @@ void EmbeddedSurfaceGenerator::initializePostSubGroups()
...
+  localIndex embSurfNodeNumberUpdated = embSurfNodeManager.size();
+  array2d< real64, nodes::REFERENCE_POSITION_PERM > & embSurfNodesPosUpdated = embSurfNodeManager.referencePosition();
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/surfaceGeneration/EmbeddedSurfaceGenerator.cpp:396]: same here.

File: src/coreComponents/physicsSolvers/surfaceGeneration/EmbeddedSurfaceGenerator.cpp [hunk 16]
```diff
@@ -109,123 +109,354 @@ void EmbeddedSurfaceGenerator::initializePostSubGroups()
...
+    using KERNEL_TYPE = decltype( kernel );
+
+    KERNEL_TYPE::template launchCIComputationKernel< parallelDevicePolicy< 32 >, KERNEL_TYPE >( kernel );
+  } );
+
 
   // add all new nodes to newObject list
-  for( localIndex ni = 0; ni < embSurfNodeManager.size(); ni++ )
+  // also, get index of edges that were just cut
+  array1d< globalIndex > newEdges;
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/surfaceGeneration/EmbeddedSurfaceGenerator.cpp:393]: you can probably use an std::vector for this. And it would be nice to initialize the size if it's known.

File: src/coreComponents/physicsSolvers/surfaceGeneration/EmbeddedSurfaceGenerator.cpp [hunk 17]
```diff
@@ -109,123 +109,354 @@ void EmbeddedSurfaceGenerator::initializePostSubGroups()
...
+  NewObjectLists newObjects; //this should probably not be re-initialized
+  // begin geometric operations
+  real64 const planeCenter[3] = LVARRAY_TENSOROPS_INIT_LOCAL_3( newFracturePlane->getCenter() );
+  real64 const normalVector[3] = LVARRAY_TENSOROPS_INIT_LOCAL_3( newFracturePlane->getNormal() );
+  // Initialize variables
+  globalIndex nodeIndex;
+  integer isPositive, isNegative;
+  real64 distVec[ 3 ];
+
+  //COPYING FROM EXISTING FUNCTION - NEEDS CLEANING
```
Comment (CusiniM) [inline src/coreComponents/physicsSolvers/surfaceGeneration/EmbeddedSurfaceGenerator.cpp:315]: can we do some cleaning. Also, I think we need to add an explanation of the algorithm. it gets hard to follow. I would also break this down in a few smaller functions to help understanding what's happening.

**Review by CusiniM — COMMENTED** [review]
First round of comments. You still have a lot of commented code etc. Please clean it up before merging and add some comments.

File: src/coreComponents/mesh/simpleGeometricObjects/BoundedPlane.cpp [hunk 18]
```diff
@@ -61,6 +61,31 @@ BoundedPlane::BoundedPlane( const string & name, Group * const parent ):
...
+  m_widthVector{ 0.0, 0.0, 0.0 },
+  m_tolerance( 1e-5 )
+{
+  m_origin = { (oldX + newX)/2.0, (oldY + newY)/2.0, 0.0};
+  m_normal = { -(newY-oldY), newX-oldX, 0.0 };
+  m_lengthVector = { newX-oldX, newY-oldY, 0.0 };
+  m_widthVector = { 0.0, 0.0, 1.0 };
+  real64 norm = std::sqrt( pow( newX-oldX, 2 )+pow( newY-oldY, 2 ));
+  m_dimensions.resize( 2 );
+  m_dimensions[0] = norm+1e-4; //small tolerance to ensure that both ends are contained in the plane - TODO: try to use m_tolerance
```
Comment (rrsettgast) [inline src/coreComponents/mesh/simpleGeometricObjects/BoundedPlane.cpp:81]: we do need to avoid hard coding tolerances.

File: src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp [hunk 19]
```diff
@@ -1144,6 +1150,25 @@ SolidMechanicsLagrangianFEM::applySystemSolution( DofManager const & dofManager,
...
+  //       localIndex a = i/3;
+  //       localIndex b = i - 3*a;
+  //       if (a<72)
+  //       {
+  //         std::cout<<"disp value: "<<totDisp(a,b)<<std::endl;
+  //       }
+  //   }
+  // }
+  //
+
```
Comment (rrsettgast) [inline src/coreComponents/physicsSolvers/solidMechanics/SolidMechanicsLagrangianFEM.cpp:1171]: remove

**Review by rrsettgast — APPROVED** [review]
